File: src/weaponsprocurement/gui/StockReviewTradePlanner.java
package weaponsprocurement.gui;

import weaponsprocurement.core.StockCategory;
import weaponsprocurement.core.StockItemType;
import weaponsprocurement.core.WeaponStockRecord;
import weaponsprocurement.core.WeaponStockSnapshot;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

final class StockReviewTradePlanner {
    private static final int MATCH_SELL = 0;
    private static final int MATCH_SELLER_BUY = 1;
    private static final int MATCH_GENERIC_BUY = 2;

    private StockReviewTradePlanner() {
    }

    static List<WeaponStockRecord> visibleTradeableRecords(WeaponStockSnapshot snapshot, StockCategory category) {
        List<WeaponStockRecord> result = new ArrayList<WeaponStockRecord>();
        if (snapshot == null || category == null) {
            return result;
        }
        addVisibleTradeableRecords(result, snapshot.getRecords(category));
        return result;
    }

    static List<WeaponStockRecord> visibleTradeableRecords(WeaponStockSnapshot snapshot, StockItemType itemType, StockCategory category) {
        List<WeaponStockRecord> result = new ArrayList<WeaponStockRecord>();
        if (snapshot == null || category == null) {
            return result;
        }
        addVisibleTradeableRecords(result, snapshot.getRecords(itemType, category));
        return result;
    }

    static List<WeaponStockRecord> visibleTradeableRecords(WeaponStockSnapshot snapshot) {
        List<WeaponStockRecord> result = new ArrayList<WeaponStockRecord>();
        if (snapshot == null) {
            return result;
        }
        for (StockCategory category : StockCategory.values()) {
            addVisibleTradeableRecords(result, snapshot.getRecords(category));
        }
        return result;
    }

    static List<WeaponStockRecord> visibleBuyableRecords(WeaponStockSnapshot snapshot) {
        List<WeaponStockRecord> result = new ArrayList<WeaponStockRecord>();
        if (snapshot == null) {
            return result;
        }
        for (StockCategory category : StockCategory.values()) {
            addVisibleBuyableRecords(result, snapshot.getRecords(category));
        }
        return result;
    }

    static List<WeaponStockRecord> cheapestFirstVisibleBuyableRecords(WeaponStockSnapshot snapshot) {
        List<WeaponStockRecord> result = visibleBuyableRecords(snapshot);
        Collections.sort(result, new CheapestBuyRecordComparator(snapshot));
        return result;
    }

    static List<StockReviewPendingPurchase> withAdjustment(List<StockReviewPendingPurchase> pendingPurchases,
                                                           String itemKey,
                                                           String submarketId,
                                                           int delta) {
        List<StockReviewPendingPurchase> result = new ArrayList<StockReviewPendingPurchase>();
        boolean adjusted = false;
        if (pendingPurchases != null) {
            for (int i = 0; i < pendingPurchases.size(); i++) {
                StockReviewPendingPurchase purchase = pendingPurchases.get(i);
                int quantity = purchase.getQuantity();
                if (purchase.matches(itemKey, submarketId)) {
                    quantity += delta;
                    adjusted = true;
                }
                if (quantity != 0) {
                    result.add(new StockReviewPendingPurchase(purchase.getItemKey(), purchase.getSubmarketId(), quantity));
                }
            }
        }
        if (!adjusted && delta != 0) {
            result.add(new StockReviewPendingPurchase(itemKey, submarketId, delta));
        }
        return result;
    }

    static List<StockReviewPendingPurchase> executionOrder(List<StockReviewPendingPurchase> pendingPurchases) {
        List<StockReviewPendingPurchase> result = new ArrayList<StockReviewPendingPurchase>();
        if (pendingPurchases == null || pendingPurchases.isEmpty()) {
            return result;
        }
        addMatching(result, pendingPurchases, MATCH_SELL);
        addMatching(result, pendingPurchases, MATCH_SELLER_BUY);
        addMatching(result, pendingPurchases, MATCH_GENERIC_BUY);
        return result;
    }

    private static void addMatching(List<StockReviewPendingPurchase> result,
                                    List<StockReviewPendingPurchase> pendingPurchases,
                                    int match) {
        for (int i = 0; i < pendingPurchases.size(); i++) {
            StockReviewPendingPurchase purchase = pendingPurchases.get(i);
            if (matches(purchase, match)) {
                result.add(purchase);
            }
        }
    }

    private static boolean matches(StockReviewPendingPurchase purchase, int match) {
        if (match == MATCH_SELL) {
            return purchase.isSell();
        }
        if (match == MATCH_SELLER_BUY) {
            return purchase.isBuy() && purchase.getSubmarketId() != null;
        }
        return purchase.isBuy() && purchase.getSubmarketId() == null;
    }

    private static void addVisibleBuyableRecords(List<WeaponStockRecord> result, List<WeaponStockRecord> records) {
        if (records == null) {
            return;
        }
        for (int i = 0; i < records.size(); i++) {
            WeaponStockRecord record = records.get(i);
            if (record.getBuyableCount() > 0) {
                result.add(record);
            }
        }
    }

    private static void addVisibleTradeableRecords(List<WeaponStockRecord> result, List<WeaponStockRecord> records) {
        if (records == null) {
            return;
        }
        for (int i = 0; i < records.size(); i++) {
            WeaponStockRecord record = records.get(i);
            if (record.getBuyableCount() > 0 || record.getPlayerCargoCount() > 0) {
                result.add(record);
            }
        }
    }

    private static final class CheapestBuyRecordComparator implements Comparator<WeaponStockRecord> {
        private final StockReviewQuoteBook quoteBook;

        CheapestBuyRecordComparator(WeaponStockSnapshot snapshot) {
            this.quoteBook = new StockReviewQuoteBook(snapshot);
        }

        @Override
        public int compare(WeaponStockRecord left, WeaponStockRecord right) {
            int result = Integer.compare(quoteBook.cheapestUnitPrice(left), quoteBook.cheapestUnitPrice(right));
            if (result != 0) {
                return result;
            }
            return left.getDisplayName().compareToIgnoreCase(right.getDisplayName());
        }
    }

}

File: src/main/kotlin/weaponsprocurement/gui/StockReviewTradePlanner.kt
package weaponsprocurement.gui

import weaponsprocurement.core.StockCategory
import weaponsprocurement.core.StockItemType
import weaponsprocurement.core.WeaponStockRecord
import weaponsprocurement.core.WeaponStockSnapshot
import java.util.ArrayList

class StockReviewTradePlanner private constructor() {
    private class CheapestBuyRecordComparator(snapshot: WeaponStockSnapshot?) : Comparator<WeaponStockRecord> {
        private val quoteBook = StockReviewQuoteBook(snapshot)

        override fun compare(left: WeaponStockRecord, right: WeaponStockRecord): Int {
            val result = quoteBook.cheapestUnitPrice(left).compareTo(quoteBook.cheapestUnitPrice(right))
            if (result != 0) {
                return result
            }
            return (left.displayName ?: "").compareTo(right.displayName ?: "", ignoreCase = true)
        }
    }

    companion object {
        private const val MATCH_SELL = 0
        private const val MATCH_SELLER_BUY = 1
        private const val MATCH_GENERIC_BUY = 2

        @JvmStatic
        fun visibleTradeableRecords(snapshot: WeaponStockSnapshot?, category: StockCategory?): List<WeaponStockRecord> {
            val result = ArrayList<WeaponStockRecord>()
            if (snapshot == null || category == null) {
                return result
            }
            addVisibleTradeableRecords(result, snapshot.getRecords(category))
            return result
        }

        @JvmStatic
        fun visibleTradeableRecords(
            snapshot: WeaponStockSnapshot?,
            itemType: StockItemType?,
            category: StockCategory?,
        ): List<WeaponStockRecord> {
            val result = ArrayList<WeaponStockRecord>()
            if (snapshot == null || category == null) {
                return result
            }
            addVisibleTradeableRecords(result, snapshot.getRecords(itemType, category))
            return result
        }

        @JvmStatic
        fun visibleTradeableRecords(snapshot: WeaponStockSnapshot?): List<WeaponStockRecord> {
            val result = ArrayList<WeaponStockRecord>()
            if (snapshot == null) {
                return result
            }
            for (category in StockCategory.values()) {
                addVisibleTradeableRecords(result, snapshot.getRecords(category))
            }
            return result
        }

        @JvmStatic
        fun visibleBuyableRecords(snapshot: WeaponStockSnapshot?): List<WeaponStockRecord> {
            val result = ArrayList<WeaponStockRecord>()
            if (snapshot == null) {
                return result
            }
            for (category in StockCategory.values()) {
                addVisibleBuyableRecords(result, snapshot.getRecords(category))
            }
            return result
        }

        @JvmStatic
        fun cheapestFirstVisibleBuyableRecords(snapshot: WeaponStockSnapshot?): List<WeaponStockRecord> {
            val result = visibleBuyableRecords(snapshot)
            return result.sortedWith(CheapestBuyRecordComparator(snapshot))
        }

        @JvmStatic
        fun withAdjustment(
            pendingTrades: List<StockReviewPendingTrade>?,
            itemKey: String?,
            submarketId: String?,
            delta: Int,
        ): List<StockReviewPendingTrade> {
            val result = ArrayList<StockReviewPendingTrade>()
            var adjusted = false
            if (pendingTrades != null) {
                for (trade in pendingTrades) {
                    var quantity = trade.quantity
                    if (trade.matches(itemKey, submarketId)) {
                        quantity += delta
                        adjusted = true
                    }
                    if (quantity != 0) {
                        addPending(result, trade.itemKey, trade.submarketId, quantity)
                    }
                }
            }
            if (!adjusted && delta != 0) {
                addPending(result, itemKey, submarketId, delta)
            }
            return result
        }

        private fun addPending(result: MutableList<StockReviewPendingTrade>, itemKey: String?, submarketId: String?, quantity: Int) {
            val trade = StockReviewPendingTrade.create(itemKey, submarketId, quantity)
            if (trade != null) {
                result.add(trade)
            }
        }

        @JvmStatic
        fun executionOrder(pendingTrades: List<StockReviewPendingTrade>?): List<StockReviewPendingTrade> {
            val result = ArrayList<StockReviewPendingTrade>()
            if (pendingTrades.isNullOrEmpty()) {
                return result
            }
            addMatching(result, pendingTrades, MATCH_SELL)
            addMatching(result, pendingTrades, MATCH_SELLER_BUY)
            addMatching(result, pendingTrades, MATCH_GENERIC_BUY)
            return result
        }

        private fun addMatching(
            result: MutableList<StockReviewPendingTrade>,
            pendingTrades: List<StockReviewPendingTrade>,
            match: Int,
        ) {
            for (trade in pendingTrades) {
                if (matches(trade, match)) {
                    result.add(trade)
                }
            }
        }

        private fun matches(trade: StockReviewPendingTrade, match: Int): Boolean {
            if (match == MATCH_SELL) {
                return trade.isSell()
            }
            if (match == MATCH_SELLER_BUY) {
                return trade.isBuy() && trade.submarketId != null
            }
            return trade.isBuy() && trade.submarketId == null
        }

        private fun addVisibleBuyableRecords(result: MutableList<WeaponStockRecord>, records: List<WeaponStockRecord>?) {
            if (records == null) {
                return
            }
            for (record in records) {
                if (record.buyableCount > 0) {
                    result.add(record)
                }
            }
        }

        private fun addVisibleTradeableRecords(result: MutableList<WeaponStockRecord>, records: List<WeaponStockRecord>?) {
            if (records == null) {
                return
            }
            for (record in records) {
                if (record.buyableCount > 0 || record.playerCargoCount > 0) {
                    result.add(record)
                }
            }
        }
    }
}
